### PR TITLE
Use new haystack views and remove the now unused search handlers

### PR DESCRIPTION
**Pull request by viggo-devries**

### CODE REVIEW

File: src/oscar/apps/search/views.py [hunk 1]
```diff
@@ -1,73 +1,121 @@
...
+            messages.error(request, _("The given page number was invalid."))
+            return redirect("catalogue:index")
+
+    def get_context_data(self, *args, **kwargs):
+        ctx = super().get_context_data(*args, **kwargs)
+        ctx["summary"] = _("All products")
+        return ctx
+
+
+class CatalogueView(BaseCatalogueView, BaseFacetedSearchView):
```
Comment (samar-hassan) [inline src/oscar/apps/search/views.py:61]: Why not move this view to apps/catalogue/views and import BaseFacetedSearchView over there.

**Review by samar-hassan — CHANGES_REQUESTED** [review]

File: src/oscar/apps/search/search_indexes.py [hunk 2]
```diff
@@ -42,15 +42,13 @@ def index_queryset(self, using=None):
...
+        return self.get_model().objects.browsable()
 
     def prepare_product_class(self, obj):
         return obj.get_product_class().name
 
     def prepare_category(self, obj):
-        categories = obj.categories.all()
-        if len(categories) > 0:
-            return [category.full_name for category in categories]
+        return [category.pk for category in obj.categories.all()] or None
```
Comment (specialunderwear) [inline src/oscar/apps/search/search_indexes.py:51]: please values_list("pk", flat=True)

File: src/oscar/apps/search/generic_views.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+class FacetedSearchView(BaseFacetedSearchView):
+    facet_fields = settings.OSCAR_SEARCH_FACETS["fields"].keys()
+    paginate_by = settings.OSCAR_PRODUCTS_PER_PAGE
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+
+        # Show suggestion no matter what.  Haystack 2.1 only shows a suggestion
+        # if there are some results, which seems a bit weird to me.
+        if self.queryset.query.backend.include_spelling:
```
Comment (specialunderwear) [inline src/oscar/apps/search/generic_views.py:20]: Please verify this is still needed.

File: src/oscar/templates/oscar/search/results.html [hunk 4]
```diff
@@ -77,11 +77,11 @@ <h4>{% trans "Refine your search" %}</h4>
         {% endif %}
     </form>
 
-    {% if page.object_list %}
+    {% if results %}
```
Comment (specialunderwear) [inline src/oscar/templates/oscar/search/results.html:80]: why no paginator anymore?

File: tests/functional/catalogue/test_catalogue.py [hunk 5]
```diff
@@ -99,15 +103,21 @@ def test_shows_pagination_navigation_for_multiple_pages(self):
         self.assertContains(page, "Page 1 of 2")
 
     def test_is_public_on(self):
-        product = create_product(upc="grote-bats", is_public=True)
+        product = create_product(
+            upc="grote-bats", is_public=True, structure="standalone"
+        )
+        call_command("rebuild_index", "--noinput")
```
Comment (specialunderwear) [inline tests/functional/catalogue/test_catalogue.py:109]: why is this needed when using "haystack.signals.RealtimeSignalProcessor"

**Review by specialunderwear — CHANGES_REQUESTED** [review]

File: tests/functional/catalogue/test_catalogue.py [hunk 6]
```diff
@@ -101,13 +105,14 @@ def test_shows_pagination_navigation_for_multiple_pages(self):
     def test_is_public_on(self):
         product = create_product(upc="grote-bats", is_public=True)
         page = self.app.get(reverse("catalogue:index"))
-        products_on_page = list(page.context["products"].all())
+        products_on_page = list(page.context["products"])
+        products_on_page = [prd.object for prd in products_on_page]
```
Comment (samar-hassan) [inline tests/functional/catalogue/test_catalogue.py:109]: `prd.object` Is this a breaking change? Is there a way to avoid this

**Review by samar-hassan — APPROVED** [review]

File: tests/functional/catalogue/test_catalogue.py (hunk 6 above)
Comment (viggo-devries) [reply tests/functional/catalogue/test_catalogue.py:109]: Yes, it's a breaking change for the frontend but right now this is just wrong.
the search results are already returned like this so it's now the same across the board.

Comment (viggo-devries) [reply tests/functional/catalogue/test_catalogue.py:109]: I will add this as an big remark in the next release note

**Review by specialunderwear — APPROVED** [review]
